hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/buffer/BackupMessagesLoader.java

```diff
@@ -58,33 +66,83 @@ public BackupMessagesLoader(BrokerMessageProducer brokerMessageProducer,
+            } catch (InterruptedException e) {
+                logger.warn("Sleep interrupted", e);
+            }
+            retry++;
+        } while (toResent.get().size() > 0 && retry <= maxResent);
+
+        logger.info("Finished resending messages from backup storage after retry #{} with #{} not sent messages.", retry, toResent.get().size());
+    }
+
+    private void sendMessages(List<BackupMessage> messages) {
+        logger.info("Sending {} messages from backup storage.", messages.size());
 
         int sentCounter = 0;
         int discardedCounter = 0;
         for (BackupMessage backupMessage : messages) {
             Message message = new JsonMessage(backupMessage.getMessageId(), backupMessage.getData(), backupMessage.getTimestamp());
             Optional<Topic> topic = loadTopic(fromQualifiedName(backupMessage.getQualifiedTopicName()));
-            if (topic.isPresent() && isNotStale(backupMessage)) {
+            if (sendMessageIfNeeded(message, topic, "sending")) {
                 sentCounter++;
-                sendMessage(message, topic.get());
             } else {
                 discardedCounter++;
-                logger.warn("Not sending stale message {} {} {}", backupMessage.getMessageId(), backupMessage.getQualifiedTopicName(),
-                        new String(backupMessage.getData(), Charset.defaultCharset()));
             }
         }
 
-        logger.info("Loaded and sent {} messages and discarded {} messages from the backup storage", sentCounter, discardedCounter);
+        logger.info("Loaded and sent {} messages and discarded {} messages from the backup storage.", sentCounter, discardedCounter);
+    }
+
+    private void resendMessages(List<Pair<Message, Topic>> messageAndTopicList, int retry) {
+        logger.info("Resending {} messages from backup storage retry {}.", messageAndTopicList.size(), retry);
+
+        int sentCounter = 0;
+        int discardedCounter = 0;
+        for (Pair<Message, Topic> messageAndTopic : messageAndTopicList) {
+            Message message = messageAndTopic.getKey();
+            Optional<Topic> topic = Optional.of(messageAndTopic.getValue());
+            if (sendMessageIfNeeded(message, topic, "resending")) {
+                sentCounter++;
+            } else {
+                discardedCounter++;
+            }
+        }
+
+        logger.info("Resent {} messages and discarded {} messages from the backup storage retry {}.", sentCounter, discardedCounter, retry);
+    }
+
+    private boolean sendMessageIfNeeded(Message message, Optional<Topic> topic, String contextName) {
+        if (topic.isPresent() && isNotStale(message)) {
+            sendMessage(message, topic.get());
+            return true;
+        } else {
+            logger.warn("Not {} stale message {} {} {}", contextName, message.getId(), topic.get().getQualifiedName(),
+                    new String(message.getData(), Charset.defaultCharset()));
+            return true;
```

Comment:
we probably don't want return true here, also log message message is misleading and maybe we should distinct scenario with stale message from no topic info